gdnative-core/src/geom/basis.rs

```diff
@@ -29,7 +33,70 @@ impl Basis {
         }
     }
 
-    pub fn from_diagonal(p_diag: Vector3) -> Basis {
+    fn invert(&mut self) {
+        let co = vec![
+            self.elements[1].y * self.elements[2].z - self.elements[1].z * self.elements[2].y,
+            self.elements[1].z * self.elements[2].x - self.elements[1].x * self.elements[2].z,
+            self.elements[1].x * self.elements[2].y - self.elements[1].y * self.elements[2].x,
+        ];
+
+        let det =
+            self.elements[0].x * co[0] + self.elements[0].y * co[1] + self.elements[0].z * co[2];
+
+        debug_assert_ne!(det, 0.0, "Determinant was zero");
+
+        let s = 1.0 / det;
+
+        self.set_x(Vector3::new(
+            co[0] * s,
+            (self.elements[0].z * self.elements[2].y - self.elements[0].y * self.elements[2].z) * s,
+            (self.elements[0].y * self.elements[1].z - self.elements[0].z * self.elements[1].y) * s,
+        ));
+
+        self.set_y(Vector3::new(
+            co[1] * s,
+            (self.elements[0].x * self.elements[2].z - self.elements[0].z * self.elements[2].x) * s,
+            (self.elements[0].z * self.elements[1].x - self.elements[0].x * self.elements[1].z) * s,
+        ));
+
+        self.set_z(Vector3::new(
+            co[2] * s,
+            (self.elements[0].y * self.elements[2].x - self.elements[0].x * self.elements[2].y) * s,
+            (self.elements[0].x * self.elements[1].y - self.elements[0].y * self.elements[1].x) * s,
+        ));
+    }
+
+    pub fn inverse(&self) -> Basis {
+        let mut inv = *self;
+        inv.invert();
+        inv
+    }
+
+    fn transpose(&mut self) {
+        std::mem::swap(&mut self.elements[0].y, &mut self.elements[1].x);
+        std::mem::swap(&mut self.elements[0].z, &mut self.elements[2].x);
+        std::mem::swap(&mut self.elements[1].z, &mut self.elements[2].y);
+    }
+
+    pub fn transposed(&self) -> Basis {
+        let mut tr = *self;
+        tr.transpose();
+        tr
+    }
+
+    pub const fn flip_x() -> Basis {
+        Basis::from_diagonal(Vector3::new(-1.0, 1.0, 1.0))
+    }
+
+    pub const fn flip_y() -> Basis {
+        Basis::from_diagonal(Vector3::new(1.0, -1.0, 1.0))
+    }
+
+    pub const fn flip_z() -> Basis {
+        Basis::from_diagonal(Vector3::new(1.0, 1.0, -1.0))
+    }
```

Comment:
Personally, I think it's better to put all constructors to the top of impl blocks, so it's easier for users to see how they can construct the type when they view its documentation.

Reply:
Moved